### ISSUE TITLE
Resolving dependencies and inheritance

### ISSUE THREAD

**Issue by sebastian-zarzycki-es** [AliSoftware/Dip#69]
I've think I've encountered somewhat unusual bug/design consequence in Dip:

```
code

class BaseModel: Resolvable {
    var baseService: BaseService!
	
	func didResolveDependencies() {
    }
}

class ChatModel : BaseModel {    
    var chatService: ChatServiceProtocol!
}

container:

container = DependencyContainer { container in
	let baseService = BaseService()
	container.register(.Singleton) { baseService as BaseService }
        
	let chatService = ChatService()
	container.register(.Singleton) { chatService as ChatServiceProtocol }
	
	let baseModel = BaseModel()
	container.register(.Singleton) { baseModel }.resolveDependencies { container, entity in
		entity.baseService = try! container.resolve() as BaseService
	}

	let chatModel = ChatModel()
	container.register(.Singleton) { chatModel }.resolveDependencies { container, entity in
		entity.chatService = try! container.resolve() as ChatServiceProtocol
	}
}
```

When I initialize ChatModel somewhere, its baseService is not injected, even though it inherits from BaseModel. 

```
let chatModel:ChatModel = ChatModel()
chatModel.chatService // ok
chatModel.baseService // nil
```

So it would seem that if objects inherit from other objects, then the properties from inherited object aren't resolved properly, because container doesn't trigger it's process in this case. My goal, was, obviously, to have a ChatModel, that would have both baseService and chatService injected, but I wanted to keep baseService in BaseModel. Did I do something wrong? Is it a bug?

I do realize, that I could initialize chatModel like so:
```
let chatModel = ChatModel()
container.register(.Singleton) { chatModel }.resolveDependencies { container, entity in
	entity.baseService = try! container.resolve() as BaseService
	entity.chatService = try! container.resolve() as ChatServiceProtocol
}
```

But that's (overall) not the point of having inheritance in the first place.

**Comment by sebastian-zarzycki-es** [comment]
Also, looking at this, I'm not exactly sure if I should register BaseService as singleton, especially if there are other objects that clearly inherit from it (so there will be multiple instances after all). But, even if, the code never reaches the creation of BaseModel

**Comment by ilyapuchka** [comment]
The problem is that container does not know anything about inheritance of your classes. Even more if you register components as implementations of abstractions it will not even know what concrete implementation you register, because 'as' effectively erase type information. 

To do what you ask for we will need to introduce something like inheritance for definitions, so that you can create base definition and create another one based on it, but that will, probably, return another instance, and for such definitions we will need to call resolve dependencies block of it's parent definitions... Looks like complicated and error prone feature that needs a good design. Typhoon btw has this feature.

If you want to remove code duplication you can achieve that by other means, like adding method `resolveServices(container: DependencyContainer)` on `BaseModel` and override it in `ChatModel`. Just put it in extension in the same file where you register all components to isolate dependency on Dip. Not the ideal solution, but that's all I can suggest for now.

Anyway thanks for reporting such things cause it really helps to see what functionality is missing.
We can keep this issue opened to discuss how/should it be implemented in Dip.

**Comment by sebastian-zarzycki-es** [comment]
Yes, I understand that this is not an easy one to solve. But if Typhoon were able to do it, then it means you can too :) For now, I accept this as a limitation of the framework, although if this is ever implemented, it would be great.

**Comment by ilyapuchka** [comment]
Looks like the only thing that makes sense to inherit from other definitions is a `resolveDependencies` block, right?

Then maybe it would be easier just to do it like this?

```swift
let baseModel = BaseModel()
let baseModelDef = container.register(.Singleton) { baseModel }
  .resolveDependencies { container, entity in
     entity.baseService = try! container.resolve() as BaseService
}

let chatModel = ChatModel()
container.register(.Singleton) { chatModel }.resolveDependencies { container, entity in
    try baseModelDef.resolveDependenciesOf(entity, withContainer: container)
    //entity.baseService = try! container.resolve() as BaseService
    entity.chatService = try! container.resolve() as ChatServiceProtocol
}
```
That method `resolveDependenciesOf` of definition is already defined, but as `internal`. We can just add it in public scope and constrain it with generic type of the instance.
With that approach user will need to make sure the correct base definition is used cause container will not store any relationships between definitions.

Another option is to add optional `parent` parameter in each `register` method:

```swift
let chatModel = ChatModel()
container.register(.Singleton, parent: baseModelDef) { chatModel }
  .resolveDependencies { container, entity in
    entity.chatService = try! container.resolve() as ChatServiceProtocol
}
```
The drawback is that it complicates `register` method that has already several parameters and overloads.

Alternative option is to use mutating method in chain:

```swift
let chatModel = ChatModel()
container.register(.Singleton) { chatModel }
  .parentDefinition(baseModelDef)
  .resolveDependencies { container, entity in
    entity.chatService = try! container.resolve() as ChatServiceProtocol
}
```

**Comment by sebastian-zarzycki-es** [comment]
None of these solutions is particularly appealing, ideally I would like to have just regular inheritance working as it always works and slap injection on top of that. If I would have to pick, though, the one with providing "parent" property is the cleanest and most in-line with inheritance.

**Comment by ilyapuchka** [comment]
Don't think that is currently possible without changing API design.
The problem is that even if we can get superclass type when instance of superclass is resolved we can not do anything with it, because we can not know if superclass implementation was ever registered in container. We could do that if while registering component we get type information about both registered abstraction (protocol) and concrete implementation (class). But we get only abstraction type (type of factory result) and personally I don't want to change that.

**Comment by ilyapuchka** [comment]
Also I'm not even sure if automatically calling `resolveDependencies` on superclass definition is a desired default behaviour.

**Comment by sebastian-zarzycki-es** [comment]
Understood. My main gripe is that you have to now keep references to registrations and reuse them later in other registrations, which is just not neat. But I'll take it anyway, if that, together with the "parent" option would allow me to keep consistent and controllable dependencies in inherited objects. Better this, than nothing, obviously ;)

**Comment by ilyapuchka** [comment]
@sebastian-zarzycki-es
Few things that we should consider to properly implement that:

1. does it make sense to inherit anything else, like factory and scope? If yes then register method with parent parameter should have all other parameters as optional
2. what should happen when parent definition is removed from the container? should childs keep weak reference to their parents or strong and keep calling parents' `resolveDependencies`?
3. we should probably make sure that when user registers child definition parent definition is registered in the same container?
4. some other edge cases you can think of?

**Comment by sebastian-zarzycki-es** [comment]
1. Not sure what you're asking about here. If that's about internals, I don't feel competent enough to answer. I would imagine that this is something you would know better how to organize. I'm not sure how big of a change, internally, is supporting inheritance in DI.
2. Is that a viable scenario? Why would you define anything in container, just to remove it later? As we've agreed on, once you call the bootstrap method, it's all sealed.
3. Of course.
4. Not really. In my mind, this should be a simple usecase, that would follow the standard inheritance rules in language. If someone tries to something silly, like removing definitions from container, then just throw an Error.

**Comment by ilyapuchka** [comment]
Looks like neither of the implementations I've proposed previously provides satisfying solution. Passing another definition as base definition in parameters does not let to ensure type-safety because of how generic works in Swift atm. Letting user to call some method on definition itself also does not provide any safety and makes api confusing.

Looks like the best way to reuse dependencies resolution from base classes is to move that responsibility from definition to class itself. We already have `Resolvable` protocol with `didResolveDependencies` method. We can add `resolveDependencies(container: DependencyContainer)` method and call it right after instance is created by container. Then base class dependencies should be resolved in this method. At the same time subclasses can override it to resolve any additional properties and simply call `super`.

The only drawback of that solution is that resolving dependencies may happen in different places.
Switching to other patterns, like constructor injection or auto-injection may solve that. #116 also contains  patch that fixes resolving inherited auto-injected properties.

**Comment by ilyapuchka** [comment]
Also addressed inheritance in #116 when using properties auto-injection
